src/DotNetty.Common/Utilities/HashedWheelTimer.cs

```diff
@@ -0,0 +1,712 @@
+        const int WorkerStateStarted = 1;
+        const int WorkerStateShutdown = 2;
+        int workerStateVolatile = WorkerStateInit; // 0 - init, 1 - started, 2 - shut down
+
+        readonly long tickDuration;
+        readonly HashedWheelBucket[] wheel;
+        readonly int mask;
+        readonly CountdownEvent startTimeInitialized = new CountdownEvent(1);
+        readonly IQueue<HashedWheelTimeout> timeouts = PlatformDependent.NewMpscQueue<HashedWheelTimeout>();
+        readonly IQueue<HashedWheelTimeout> cancelledTimeouts = PlatformDependent.NewMpscQueue<HashedWheelTimeout>();
+        internal long PendingTimeouts;
+        readonly long maxPendingTimeouts;
+        long startTimeVolatile;
+
+        public HashedWheelTimer()
+            : this(TimeSpan.FromMilliseconds(100), 512, -1)
+        {
+        }
+
+        /// <summary>
+        /// Creates a new timer.
+        /// </summary>
+        /// <param name="tickDuration">the duration between tick</param>
+        /// <param name="ticksPerWheel">the size of the wheel</param>
+        /// <param name="maxPendingTimeouts">The maximum number of pending timeouts after which call to
+        /// <c>newTimeout</c> will result in <see cref="RejectedExecutionException"/> being thrown.
+        /// No maximum pending timeouts limit is assumed if this value is 0 or negative.</param>
+        /// <exception cref="ArgumentException">if either of <c>tickDuration</c> and <c>ticksPerWheel</c> is &lt;= 0</exception>
+        public HashedWheelTimer(
+            TimeSpan tickDuration,
+            int ticksPerWheel,
+            long maxPendingTimeouts)
+        {
+            if (tickDuration <= TimeSpan.Zero)
+            {
+                throw new ArgumentException("tickDuration must be greater than 0: " + tickDuration);
+            }
+            if (Math.Ceiling(tickDuration.TotalMilliseconds) > int.MaxValue)
+            {
+                throw new ArgumentException($"{nameof(tickDuration)} must be less than or equal to ${int.MaxValue} ms.");
+            }
+            if (ticksPerWheel <= 0)
+            {
+                throw new ArgumentException("ticksPerWheel must be greater than 0: " + ticksPerWheel);
+            }
+
+            // Normalize ticksPerWheel to power of two and initialize the wheel.
+            this.wheel = CreateWheel(ticksPerWheel);
+            this.worker = new Worker(this);
+            this.mask = this.wheel.Length - 1;
+
+            this.tickDuration = tickDuration.Ticks;
+
+            // Prevent overflow
+            if (this.tickDuration >= long.MaxValue / this.wheel.Length)
+            {
+                throw new ArgumentException(
+                    string.Format(
+                        "tickDuration: {0} (expected: 0 < tickDuration in nanos < {1}",
+                        tickDuration,
+                        long.MaxValue / this.wheel.Length));
+            }
+            this.workerThread = new XThread(st => this.worker.Run());
```

Comment:
yes, thread won't start until later use of the HWT object itself which cannot happen before the ctor is completed.

---
In reply to: [114473937](https://github.com/Azure/DotNetty/pull/242#discussion_r114473937) [](ancestors = 114473937)